### ISSUE TITLE
Collaboration Opportunity: Join Forces for Duplicating GitHub Action Work

### ISSUE THREAD

**Issue by q-uint** [dfinity/setup-dfx#1]
I just came across this new code repository, which appears to be aimed at creating a GitHub Action to install `dfx`, much like the one our team has already developed two years ago.

I believe there might be a fantastic opportunity for us to collaborate and avoid unnecessary duplication of efforts. Since our goals align with enhancing the developer experience and promoting the adoption of `dfx`, I thought it would be beneficial for us to explore working together.

Rather than proceeding with separate development paths, I propose that we consider the following options:

1. Collaborative Development: Join forces and contribute to a single GitHub Action repository. This way, we can pool our expertise and resources to create an even more robust and feature-rich solution.

2. Integration: If your team has specific requirements or enhancements in mind, we can discuss integrating those into our existing GitHub Action. Likewise, we are open to contributing any valuable functionalities we've developed to your repository.

3. Shared Maintenance: We can share the maintenance responsibilities, ensuring that updates, bug fixes, and improvements are managed efficiently.

By combining our efforts, we can not only save valuable development time but also foster a stronger community around `dfx`, benefitting the larger developer community.

I would be delighted to arrange a meeting to discuss this further and explore how we can collaborate effectively. Please let me know your thoughts on this matter, and if you are open to exploring collaboration possibilities.

---

https://github.com/aviate-labs/setup-dfx

**Comment by q-uint** [comment]
Hey @smallstepman, thank you for the prompt response!

---

Some reasons why using TypeScript for this GitHub Action is beneficial (subjective):

1. **Developer-Friendly**: TypeScript provides better tooling and IDE support, making it easier for developers to write, read, and debug code. It offers features like code completion, static type checking, and intelligent code suggestions, which can improve productivity and reduce errors.

2. **Scalability**: As your project evolves, TypeScript can help handle increased complexity without sacrificing readability. The type system allows catching potential issues early on and helps maintain a coherent codebase.

3. **Community and Support**: TypeScript has a large and active community, which means more resources, libraries, and tools available for integration. We have had multiple contributions from the community, I think bash would inhibit this for some people.

4. **Documentation and Readability**: TypeScript encourages writing clear and self-explanatory code. With well-defined types and interfaces, the codebase becomes more self-documenting, making it easier for new contributors to understand and get onboarded.

---

Regarding the features of the GitHub Action:

- **DFX/VESSEL/MOC Integration**: The action facilitates seamless integration with DFX, Vessel, and MOC, making it easy to set up and use these tools within workflows.

- **Envvar Support**: The action provides support for environment variables, allowing users to pass custom configurations or secrets securely.

- **Specific DFX Version Installation**: It allows users to specify a particular version of 
DFX, ensuring consistent and reproducible builds.

---

In conclusion, using TypeScript in combination with the provided features enhances the action's maintainability, developer experience, and extensibility, while also ensuring better long-term support and community adoption.

Loading the GH Actions takes less than a second, so I am no that worried about performance. The `main.ts` file is less than 100 LoC.

Let me know if you have any other questions or if there's anything specific you'd like me to look into!

**Comment by smallstepman** [comment]
I appreciate the time you took to write this. This is my personal perspective, but I find the pure bash solution to be more straightforward, easier to manage, and easier to read, especially for tasks like this (dependency installation). 

Bash also has several features that make it easy to work with (`set -x`, `shellcheck` linter), `Developer-Friendly` is a subjective notion; after all, someone may find working with `C` as the most developer-friendly option simply because this is what they're familiar with; in the end, it's all about familiarity. 

Regarding point `2. Scalability`, for the time being, we currently don't expect this GHA to get any more complex than [this](https://github.com/mmicu/sdk/blob/630ed328bdb69a82adfee2fe51b6bc4b097c275f/action.yml).

Regarding point `3. Community and Support` I could see that there are more JS/TS devs in the world; however, the owner of `setup-dfx` is ultimately @dfinity/sdk, and most of us are more familiar with bash than with TS/JS and the plethora of tools that come with this ecosystem.

Regarding point `4. Documentation and Readability`, I think the above-linked file is short enough to be easily readable and doesn't require extensive documentation.

Regarding features:
- `DFX installation` - covered by yml-bash script linked above 
- `VESSEL/MOC Integration` - we'll probably compartmentalize and introduce another GHA repo `dfinity/setup-motoko-toolchain`; thank you for the inspiration!
- `Envvar Support` - supported natively by `run` in GHA yamls:
```yml
- run: <some-cmd>
  env: 
    ENVVAR1: XYZ
    ENVVAR2: ABC
```
- `Specific DFX Version Installation` - covered by `env:`, example: https://github.com/smallstepman/eseses/blob/d11773dcc63da582daea95056e946d6374f85c77/.github/workflows/test.yml and also (since controlling the version is very important) it has been exposed via `with: ` clasue like here https://github.com/smallstepman/eseses/blob/388d01dad895d560d2d8e5eb7bf884f7e8557858/.github/workflows/test.yml#L14-L17

I'll be looking toward other opportunities for us to collaborate! For now, I'm closing this issue, but feel free to reopen it if you have other ideas you'd like us to consider!
